### PR TITLE
Send messages to remotes

### PR DESCRIPTION
**Pull request by pcapriotti**
This propagates messages to all participants of a conversation, including remote ones. Note that we still cannot handle message sending to conversations in other domains, and attempting to do so still results in a `federation-not-implemented` error.

## Checklist

 - [x] Title of this PR explains the impact of the change.
 - [x] The description provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] The CHANGELOG.md file in the *Unreleased* section has been updated to explain the change which will be included in the release notes.

### CODE REVIEW

File: libs/wire-api/src/Wire/API/Message.hs [hunk 1]
```diff
@@ -206,6 +210,78 @@ protolensToQualifiedNewOtrMessage protoMsg = do
 protolensToClientId :: Proto.Otr.ClientId -> ClientId
 protolensToClientId = newClientId . view Proto.Otr.client
 
+--- functions to generate protobuf messages
+
+mkQualifiedOtrPayload :: ClientId -> [(Qualified UserId, ClientId, ByteString)] -> ByteString -> ClientMismatchStrategy -> Proto.Otr.QualifiedNewOtrMessage
+mkQualifiedOtrPayload sender recipients dat strat =
```
Comment (jschaul) [inline libs/wire-api/src/Wire/API/Message.hs:216]: Could these functions here have a unit round trip test? Given a few haskell values, make a proto OTR payload, then decode them again and expect the same?
(I'm fine to add that test in a follow-up PR also; mostly to guard against future regressions)

File: services/galley/src/Galley/API/Message.hs [hunk 2]
```diff
@@ -213,26 +220,110 @@ postQualifiedOtrMessage senderType sender mconn convId msg = runUnionT $ do
...
+  fmap (setOf ((itraversed <. folded) . withIndex . to (\(d, (u, c)) -> (d, u, c)))) . flip Map.traverseWithKey messageMap $ \d msgs ->
+    if d == localDomain
+      then do
+        sendLocalMessages now sender senderClient mconn conv localMemberMap metadata msgs
+        pure mempty
+      else do
+        x <- sendRemoteMessages d now sender senderClient conv metadata msgs
+        case x of
+          Nothing -> pure mempty
+          Just e -> do
```
Comment (jschaul) [inline services/galley/src/Galley/API/Message.hs:263]: This code could be refactored a little for improved readability using smaller named helper functions. the lines 253 and 254 look a little bit like they partition the map of domain/user/clients into local and remote, so perhaps a helper function could be created here?

File: services/galley/src/Galley/API/Message.hs [hunk 3]
```diff
@@ -213,26 +220,110 @@ postQualifiedOtrMessage senderType sender mconn convId msg = runUnionT $ do
...
+            FederatedGalley.rmData = Just (toBase64Text (mmData metadata)),
+            FederatedGalley.rmSender = sender,
+            FederatedGalley.rmSenderClient = senderClient,
+            FederatedGalley.rmConversation = conv,
+            FederatedGalley.rmPriority = mmNativePriority metadata,
+            FederatedGalley.rmPush = mmNativePush metadata,
+            FederatedGalley.rmTransient = mmTransient metadata,
+            FederatedGalley.rmRecipients = UserClientMap rcpts
+          }
+  -- TODO: we should not need to pass the local domain to the RPC
```
Comment (jschaul) [inline services/galley/src/Galley/API/Message.hs:324]:
```suggestion
  -- FUTUREWORK: we should not need to pass the local domain to the RPC
```

File: libs/wire-api/src/Wire/API/Message.hs (hunk 1 above)
Comment (jschaul) [reply libs/wire-api/src/Wire/API/Message.hs:216]: Update: Ah I see these functions were moved from the tests into wire-api. Still, some roundtrip tests might be nice as a FUTUREWORK.

**Review by jschaul — APPROVED** [review]
🎆

File: services/galley/src/Galley/API/Message.hs (hunk 2 above)
Comment (pcapriotti) [reply services/galley/src/Galley/API/Message.hs:263]: I moved the local/remote distinction logic to a separate function with a guard, and inline the remote error handling in the `sendMessagesRemote` function.

I've also used the opportunity to get rid of the extra flattening/unflattening of `failedToSend`.

Hopefully things are a bit clearer now.

File: libs/wire-api/src/Wire/API/Message.hs (hunk 1 above)
Comment (pcapriotti) [reply libs/wire-api/src/Wire/API/Message.hs:217]: Yes, definitely. I added a FUTUREWORK.

File: services/galley/src/Galley/API/Message.hs (hunk 2 above)
Comment (jschaul) [reply services/galley/src/Galley/API/Message.hs]: Thank you! Looks more readable now.

**Pull request merged**